agenthub/codeact_agent/codeact_agent.py

```diff
@@ -211,6 +214,16 @@ def _get_messages(self, state: State) -> list[dict[str, str]]:
             {'role': 'user', 'content': self.in_context_example},
         ]
 
+        workspace_contents = ', '.join(os.listdir(config.workspace_base))
+        if not state.history:
```

Comment:
I'd prefer not to do have it implemented this way, as 
- it circumvents the whole file filtering logic we have in `opendevin/server/listen.py`
- there could be hundreds of files already there and we'd potentially waste a lot of tokens

Please extend this PR like this:
- Refactor the main functionality of method `list_files` into a general-use method with 2 parameters `path` (string) and `max_files` (number of max files to return) parameters.
- This new method could be put into a (new) .py file in `opendevin/runtime/utils`
- Adapt list_files in listen.py accordingly to use this new method
- Import that new file here, call that method accordingly; maybe use a limit of 100 for the 2nd params here.

Other option, in order to not spam the `messages` here, we could just add a simple text message here as to "The workplace folder already contains files. Use cat to list relevant files for your task." or similar in meaning.

Reply:
> * whole file filtering logic

could you add the permalink?


> * Import that new file here, call that method accordingly; maybe use a limit of 100 for the 2nd params here.

This is not a recursive list—just 32 items in our project.



> Use cat to list relevant files for your task

it will display the file content. `ls` will do that. but that's the waste of tokens; in the current method, if the prompt is 500 words + workspace content 100 words = 600; in your method, prompt 500 words + response 50 + 2nd prompt 500+50+50 = 550 + 600

Reply:
Fixed the cat with ls and added the permalink in my above comment.

> This is not a recursive list—just 32 items in our project.

We do not know how many files users will have in their `workspace` and we shouldn't assume to know if it's less or more than wherever the 32 comes from. It's about having a way to put a limit to the amount of filenames sent.

In my _alternative_ suggested method: 
it would give the model the _option_ to apply its own file filter/mask: if the user's prompt is about typescript, it could choose to only list files with .ts/.tsx extentions etc.

As to you trying to count tokens and making this an argument everytime:
who says that the first time this method is called is actually the right folder where it needs to work in?
The list of files/folders may not even give the model any help for its task if the actual target base of files is located in a folder named for some user's project. Then it'd be a waste of all sent tokens.

Reply:
Also, please going forward add at least one sentence in your issue/PR description, not only a single reference.
We expect that from every contributed PR. No discussion needed about this.

Reply:
Kinda like duplicating the title.

Reply:
I understand the concern about token usage. However, providing a comprehensive view of the workspace is essential for flexibility and completeness in various tasks. For example, if a user is working on a task involving both TypeScript and JSON configuration files, limiting the file types could miss critical context provided by the JSON files. This holistic approach ensures a better user experience and thoroughness in handling diverse tasks, despite the additional token usage.

Reply:
That's what your AI may think, but it ignores the fact that my comment was an example, not a fixed message.

> Kinda like duplicating the title.

If it is sufficient explanation of what the PR does and how, then yes.

Reply:
Replied to your example. Do you have any other examples that violate this reply?

Reply:
Instead of discussing the suggested option, which you don't agree with, focus more on the refactor.